pubsub.go

```diff
@@ -999,59 +991,66 @@ func (p *PubSub) pushMsg(msg *Message) {
 		return
 	}
 
+	err := p.checkSignature(msg)
+	if err != nil {
+		log.Debugf("dropping message from %s: %s", src, err)
+		return
+	}
+
+	// reject messages claiming to be from ourselves but not locally published
+	self := p.host.ID()
+	if peer.ID(msg.GetFrom()) == self && src != self {
+		log.Debugf("dropping message claiming to be from self but forwarded from %s", src)
+		p.tracer.RejectMessage(msg, RejectSelfOrigin)
+		return
+	}
+
+	// have we already seen and validated this message?
+	id := p.msgID(msg.Message)
+	if p.seenMessage(id) {
+		p.tracer.DuplicateMessage(msg)
+		return
+	}
```

Comment:
the default path does it this way so that an attacker can't squat the (default) message ID of a target by guessing its seqno.

Reply:
We can:

1. Check if we've already seen the message. It doesn't really matter if the signature is valid or not here. At worst, we reject a duplicate message with an invalid signature.
2. Validate the signature.
3. Mark that we've seen the message (aborting if some parallel process also saw the message).

As far as I can tell, that's what we did before as well.

Reply:
sure, we can add an early read only check before checking the signature.

Reply:
so here the signature check is consistency wrt to our signing policy; it doesn't actually verify the signature just yet. I will rename this method for clarity and also add an early read-only check before we verify the signature, since we can skip it for duplicates.

Reply:
ah wait, we already have the read only check; I will simply rename the `checktSignature` thing to `checkSigningPolicy`.

Reply:
done, renamed `checkSignature` to `checkSigningPolicy` to avoid confusion.

Here is a summary of how it works:
- when an incoming message is received (or published), we first check the signing policy for consistent presence of signatures; if it is not consistent, then it is rejected (we don't actually verify the signature in this step).
- we then perform a read only check in the seenMessages cache using the message id. If we have already marked the message, it is dropped as duplicate
- we validate the signature next. If signature verification fails, we reject the message -- not that we don't mark the message as seen in this step, as this would open an avenue for attack with the default signing policy.
- once signature validation passes, we mark the message as seen; if it was already marked (potentially because there is concurrent validation in progress), we drop it as duplicate.
- we then apply user validators, which happens exactly once for a given message.